File: internal/gitkit/ts/tests/tdd-classify.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { skipCategory } from "../bin/tdd-classify.ts";

test("DB migrations are skip:migration", () => {
  assert.equal(skipCategory("apps/api/migrations/1718900000_add_col.ts"), "migration");
  assert.equal(skipCategory("prisma/migrations/20240101_init/migration.sql"), "migration");
  assert.equal(skipCategory("packages/db/src/migration/0007-foo.ts"), "migration");
});

test("lockfiles are skip:deps (but package.json is not)", () => {
  assert.equal(skipCategory("pnpm-lock.yaml"), "deps");
  assert.equal(skipCategory("apps/web/package-lock.json"), "deps");
  assert.equal(skipCategory("bun.lock"), "deps");
  assert.equal(skipCategory("go.sum"), "deps");
  assert.equal(skipCategory("package.json"), null); // a version bump here still gets judgment
});

test("CI/CD config is skip:ci (and wins over generic yaml)", () => {
  assert.equal(skipCategory(".github/workflows/ci.yml"), "ci");
  assert.equal(skipCategory(".gitlab-ci.yml"), "ci");
  assert.equal(skipCategory(".circleci/config.yml"), "ci");
});

test("infra-as-code is skip:iac", () => {
  assert.equal(skipCategory("Dockerfile"), "iac");
  assert.equal(skipCategory("apps/api/Dockerfile.prod"), "iac");
  assert.equal(skipCategory("docker-compose.yml"), "iac");
  assert.equal(skipCategory("infra/main.tf"), "iac");
  assert.equal(skipCategory("nginx/site.conf"), "iac");
  assert.equal(skipCategory(".env.production"), "iac");
});

test("generated / build output is skip:generated", () => {
  assert.equal(skipCategory("src/api/types.generated.ts"), "generated");
  assert.equal(skipCategory("dist/index.js"), "generated");
  assert.equal(skipCategory("apps/web/src/orval.d.ts"), "generated");
  assert.equal(skipCategory("src/__generated__/schema.ts"), "generated");
});

test("docs/text files are skip:docs", () => {
  assert.equal(skipCategory("README.md"), "docs");
  assert.equal(skipCategory("docs/guide.mdx"), "docs");
  assert.equal(skipCategory("CHANGELOG.txt"), "docs");
});

test("real source code returns null (→ apply the behavioral gates)", () => {
  assert.equal(skipCategory("apps/api/src/offer.service.ts"), null);
  assert.equal(skipCategory("src/utils/money.ts"), null);
  assert.equal(skipCategory("packages/core/lib/rateLimit.ts"), null);
});

File: internal/gitkit/tddclassify_test.go
package gitkit

import (
	"bytes"
	"testing"
)

func TestSkipCategory(t *testing.T) {
	cases := map[string][]string{
		// DB migrations
		"migration": {"apps/api/migrations/1718900000_add_col.ts", "prisma/migrations/20240101_init/migration.sql", "packages/db/src/migration/0007-foo.ts"},
		// lockfiles (package.json is not — a version bump there still gets judgment)
		"deps": {"pnpm-lock.yaml", "apps/web/package-lock.json", "bun.lock", "go.sum"},
		// CI/CD config wins over generic yaml
		"ci":        {".github/workflows/ci.yml", ".gitlab-ci.yml", ".circleci/config.yml"},
		"iac":       {"Dockerfile", "apps/api/Dockerfile.prod", "docker-compose.yml", "infra/main.tf", "nginx/site.conf", ".env.production"},
		"generated": {"src/api/types.generated.ts", "dist/index.js", "apps/web/src/orval.d.ts", "src/__generated__/schema.ts"},
		"docs":      {"README.md", "docs/guide.mdx", "CHANGELOG.txt"},
		// real source → apply the behavioral gates
		"": {"package.json", "apps/api/src/offer.service.ts", "src/utils/money.ts", "packages/core/lib/rateLimit.ts"},
	}
	for want, paths := range cases {
		for _, path := range paths {
			if got := SkipCategory(path); got != want {
				t.Errorf("SkipCategory(%q) = %q, want %q", path, got, want)
			}
		}
	}
}

func TestTDDClassifyVerb(t *testing.T) {
	for _, tc := range []struct {
		args           []string
		code           int
		stdout, stderr string
	}{
		{[]string{"--json", "go.sum"}, 0, "deps\n", ""},
		{[]string{"src/a.ts"}, 0, "null\n", ""},
		{nil, 1, "", "error: usage: tdd-classify.ts <path>\n"},
	} {
		var stdout, stderr bytes.Buffer
		if code := runTDDClassify(tc.args, &stdout, &stderr); code != tc.code || stdout.String() != tc.stdout || stderr.String() != tc.stderr {
			t.Errorf("%v → %d %q %q; want %d %q %q", tc.args, code, stdout.String(), stderr.String(), tc.code, tc.stdout, tc.stderr)
		}
	}
}
